### ISSUE TITLE
Decouple kubeadm phases make make each phase idempotent

### ISSUE THREAD

**Issue by chendave** [kubernetes/kubeadm#2769]
## Is this a BUG REPORT or FEATURE REQUEST?

FEATURE REQUEST

Kubeadm now has broken down each command to sub-phases,  we should try to decouple of those phases with others and make sure each phases could be run as a independent steps.

The benefit is that, 
- If a monolithic command, such as `kubeadm init...` failed at a specific phase, we can fix it and run this phase instead of `reset` and run the whole command again.
- Inspect much more details for what each phase does, it could be helpful for us to understand what's going to do for this specific phase or get more insight for debugging purpose.

I am hoping this issue could be an umbrella issue, and something of my mind by far is,

* [ ] We need to pass down the `dry-run` option to each phases properly, e.g. https://github.com/kubernetes/kubernetes/pull/112945 and https://github.com/kubernetes/kubernetes/pull/113005.

* [ ] Sanity check of each phase and make sure there is no stale data left, e.g. https://github.com/kubernetes/kubernetes/pull/110972

* [ ] Identify the root cause for why the phase cannot be run on it's own and try to fix it is possible.

* [ ] Make sure run phases cover all the functionalities which is only meant for a  monolithic command
Something related is that `kubeadm certs renew ...` doesn't support to renew the certs located other than the default folder.

/cc @neolit123 @SataQiu @pacoxu

**Comment by neolit123** [comment]
i think we had a similar discussion about reset if i am not mistaken.

currently most phases are more of a sequence than independent actions. some are indempotent already. we could use the new Dependencies field of phases, to better chain them together too.

making everything indempotent might be doable yet we might face cases where we actually don't want a phase to be indempotent.

i am not necessarily against this change as long as we don't break user expectations, given the CLI is GA.

**Comment by chendave** [comment]
> making everything indempotent might be doable yet we might face cases where we actually don't want a phase to be indempotent.

agreed, this won't be for every phases, we need some basic analysis and get some consensus if we reach to some points.

**Comment by chendave** [comment]
Have evaluated most of command, just a few left, I will try to finished them in this week.

**Comment by chendave** [comment]
Finished check all the phases of `reset`, `join` and `init`, here is some just minor issues found, 

`kubeadm init`

- there is no way to run `show-join-command` on its own, as it has the dependencies over `bootstrap-token`, `upload-certs`, the workaround for the minimal phases to be executed will look like below, 
```kubeadm init --skip-phases=preflight,certs,kubeconfig,kubelet-start,control-plane,etcd,wait-control-plane,upload-config,mark-control-plane,kubelet-finalize,addon```

not sure if the sub-command `phase` should accept a list instead, then it will be nicer to just run:

```kubeadm init phases=bootstrap-token,upload-certs,show-join-command```


- `kubeadm init phase kubelet-start` won't throw any errors if `kubelet` is failed to start, the error msg could only be inspected via the log of the kubelet. I think it would be great if `kubeadm` can tell more about the status of the kubelet service.


`kubeadm join`

- `kubeadm join phase check-etcd` cannot run on its own, need to inherit the flag from `--control-plane` flag from it's parent, since it's a hidden phase and won't expose to end-user, so I think it is fine.

**Comment by chendave** [comment]
We can close this now.

/close
